src/Auth0.php

```diff
@@ -2,10 +2,395 @@
+    }
+    if (isset($config['guzzle_options'])) {
+        $this->guzzleOptions = $config['guzzle_options'];
+    }
+    if (isset($config['debug'])) {
+        $this->debug_mode = $config['debug'];
+    } else {
+        $this->debug_mode = false;
+    }
+    // User info is persisted unless said otherwise
+    if (isset($config['persist_user']) && $config['persist_user'] === false) {
+        $this->dontPersist('user');
+    }
+    // Access token is not persisted unless said otherwise
+    if (!isset($config['persist_access_token']) || (isset($config['persist_access_token']) &&
+            $config['persist_access_token'] === false)) {
+        $this->dontPersist('access_token');
+    }
+    // Refresh token is not persisted unless said otherwise
+    if (!isset($config['persist_refresh_token']) || (isset($config['persist_refresh_token']) &&
+            $config['persist_refresh_token'] === false)) {
+        $this->dontPersist('refresh_token');
+    }
+    // Id token is not per persisted unless said otherwise
+    if (!isset($config['persist_id_token']) || (isset($config['persist_id_token']) &&
+            $config['persist_id_token'] === false)) {
+        $this->dontPersist('id_token');
+    }
+    if (isset($config['store'])) {
+        if ($config['store'] === false) {
+            $this->store = new EmptyStore();
+        } else {
+            $this->store = $config['store'];
+        }
+    } else {
+        $this->store = new SessionStore();
+    }
+
+    $this->authentication = new Authentication ($this->domain, $this->client_id, $this->client_secret, $this->guzzleOptions);
+
+    $this->user = $this->store->get("user");
+    $this->access_token = $this->store->get("access_token");
+    $this->id_token = $this->store->get("id_token");
+    $this->refresh_token = $this->store->get("refresh_token");
+  }
+
+  public function login($state = null, $connection = null) {
+    $params = [];
+    if ($this->audience) {
+      $params['audience'] = $this->audience;
+    }
+    if ($this->scope) {
+      $params['scope'] = $this->scope;
+    }
+
+    $params['response_mode'] = $this->response_mode;
+
+    $url = $this->authentication->get_authorize_link($this->response_type, $this->redirect_uri, $connection, $state, $params);
+
+    header("Location: $url");
+    exit;
+  }
+
+  public function getUser() {
+    $this->exchange();
+    return $this->user;
+  }
+
+  public function getIdToken() {
+    $this->exchange();
+    return $this->id_token;
+  }
+
+  public function getAccessToken() {
+    $this->exchange();
+    return $this->access_token;
+  }
+
+  public function getRefreshToken() {
+    $this->exchange();
+    return $this->refresh_token;
+  }
+
+  /**
+   * Code exchange
+   *
+   * @throws CoreException If there is an active session already.
+   */
+  public function exchange() {
+    $code = $this->getAuthorizationCode();
+    if (!$code) {
+      return false;
+    }
+
+    if ($this->user) {
```

Comment:
mmm no, because if there is no `code` present it wil not do the exchange

```php
    if (!$code) {
       return false;
     }```

anyway, maybe we should add a parameter to disable token exchange? (and avoid any other `code` parameter trigger that code)